### PR TITLE
feat(ui): Add ability to select multiple workflows from list and perform actions on them. Fixes #3185

### LINKED ISSUE #3185: [UI] Perform batch action on workflows
In the same spirit as in https://github.com/kubernetes/dashboard/issues/5239, it would be nice to be able to select multiple workflows and perform batch action on it from the web UI.

### PR DESCRIPTION
**Pull request by rbreeze**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: ui/src/app/shared/services/workflows-service.ts [hunk 1]
```diff
@@ -31,7 +31,8 @@ export class WorkflowsService {
             'items.metadata.labels',
             'items.status.phase',
             'items.status.finishedAt',
-            'items.status.startedAt'
+            'items.status.startedAt',
+            'items.spec'
```
Comment (alexec) [inline ui/src/app/shared/services/workflows-service.ts:35]: this is a lot of data - can we avoid it?

Comment (rbreeze) [reply ui/src/app/shared/services/workflows-service.ts:35]: I'll check if it can be further specified. I believe I added that for the Util functions for checking to see if an action can be performed on a wf. I'll also see if that can be determined without using `spec` at all.

Comment (rbreeze) [reply ui/src/app/shared/services/workflows-service.ts:35]: Only needed `item.spec.suspended` for the check. Just fixed

File: ui/src/app/shared/workflow-actions.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+export type ActionDisabled = {
+    [action in WorkflowActionName]: boolean;
+};
+
+export type WorkflowActionName = 'retry' | 'resubmit' | 'suspend' | 'resume' | 'stop' | 'terminate' | 'delete';
+
+export interface WorkflowAction {
+    title: string;
+    action: () => Promise<any>;
```
Comment (simster7) [inline ui/src/app/shared/workflow-actions.ts:13]: What does this `Promise` return? Does it have to be `any`?

File: ui/src/app/workflows/components/workflow-details/workflow-details.tsx [hunk 3]
```diff
@@ -177,159 +177,56 @@ export class WorkflowDetails extends React.Component<RouteComponentProps<any>, W
         );
     }
 
+    private confirmAction(title: string): void {
+        if (!confirm(`Are you sure you want to ${title} this workflow?`)) {
+            return;
+        }
+        return;
+    }
```
Comment (simster7) [inline ui/src/app/workflows/components/workflow-details/workflow-details.tsx:185]: Is this function correct? It doesn't seem the `if` statement has any effect here

File: ui/src/app/workflows/components/workflow-details/workflow-details.tsx [hunk 4]
```diff
@@ -177,159 +177,56 @@ export class WorkflowDetails extends React.Component<RouteComponentProps<any>, W
...
-                action: () => this.terminateWorkflow(ctx)
-            },
-            {
-                title: 'Delete',
-                iconClassName: 'fa fa-trash',
-                action: () => this.deleteWorkflow(ctx)
-            }
-        ];
+        const actions: any = Actions.WorkflowActions;
+        const items = Object.keys(actions).map(actionName => {
```
Comment (simster7) [inline ui/src/app/workflows/components/workflow-details/workflow-details.tsx:205]: Are we sure that this returns the items in the same order every time? We don't want to have the buttons shuffle around.

File: ui/src/app/workflows/components/workflow-details/workflow-details.tsx [hunk 5]
```diff
@@ -177,159 +177,56 @@ export class WorkflowDetails extends React.Component<RouteComponentProps<any>, W
...
+            const action = actions[actionName];
+            return {
+                title: action.title.charAt(0).toUpperCase() + action.title.slice(1),
```
Comment (simster7) [inline ui/src/app/workflows/components/workflow-details/workflow-details.tsx:208]: This seems a tad dirty. Is there a reason as to why `action.action` is lower case? Could it be capital case instead and if it needs to be lower case then we can do `action.action.lower()` or whatever the TS equivalent is

File: ui/src/app/workflows/components/workflow-details/workflow-details.tsx [hunk 6]
```diff
@@ -177,159 +177,56 @@ export class WorkflowDetails extends React.Component<RouteComponentProps<any>, W
...
+    }
+
+    private getHandleErrorFunction(title: string): () => void {
+        return () => {
+            this.appContext.apis.notifications.show({
+                content: `Unable to ${title} workflow`,
+                type: NotificationType.Error
+            });
+        };
+    }
```
Comment (simster7) [inline ui/src/app/workflows/components/workflow-details/workflow-details.tsx:194]: This can probably just be inlined

File: ui/src/app/shared/workflow-actions.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    iconClassName: string;
+    disabled: (wf: Workflow) => boolean;
```
Comment (simster7) [inline ui/src/app/shared/workflow-actions.ts:15]:
```suggestion
    isDisabled: (wf: Workflow) => boolean;
```
is more clear

File: ui/src/app/workflows/components/workflows-list/workflows-list.tsx [hunk 8]
```diff
@@ -33,10 +35,22 @@ interface State {
     namespace: string;
     selectedPhases: string[];
     selectedLabels: string[];
+    selectedWorkflows: {[index: string]: models.Workflow};
     workflows?: Workflow[];
     error?: Error;
+    batchActionDisabled: Actions.ActionDisabled;
 }
 
+const allBatchActionsEnabled = {
```
Comment (simster7) [inline ui/src/app/workflows/components/workflows-list/workflows-list.tsx:44]: This could use a type

File: ui/src/app/workflows/components/workflows-list/workflows-list.tsx [hunk 9]
```diff
@@ -53,15 +67,19 @@ export class WorkflowsList extends BasePage<RouteComponentProps<any>, State> {
...
-            selectedLabels: this.queryParams('label')
+            selectedLabels: this.queryParams('label'),
+            selectedWorkflows: {},
+            batchActionDisabled: {...allBatchActionsEnabled}
         };
     }
 
     public componentDidMount(): void {
         this.fetchWorkflows(this.state.namespace, this.state.selectedPhases, this.state.selectedLabels, this.state.pagination);
+        this.setState({selectedWorkflows: {}});
```
Comment (simster7) [inline ui/src/app/workflows/components/workflows-list/workflows-list.tsx:78]: Why do we need to set this here? Why is it not enough to just do so in the constructor?

File: ui/src/app/workflows/components/workflows-list/workflows-list.tsx [hunk 10]
```diff
@@ -53,15 +67,19 @@ export class WorkflowsList extends BasePage<RouteComponentProps<any>, State> {
...
     }
 
     public componentWillUnmount(): void {
+        this.setState({selectedWorkflows: {}});
```
Comment (simster7) [inline ui/src/app/workflows/components/workflows-list/workflows-list.tsx:82]: Same here

File: ui/src/app/workflows/components/workflows-list/workflows-list.tsx [hunk 11]
```diff
@@ -159,12 +185,14 @@ export class WorkflowsList extends BasePage<RouteComponentProps<any>, State> {
         }
         workflowList
             .then(wfList => {
+                const selectedWorkflows: {[index: string]: models.Workflow} = {};
```
Comment (simster7) [inline ui/src/app/workflows/components/workflows-list/workflows-list.tsx:188]: Can this just be inlined?

File: ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+require('./workflows-toolbar.scss');
+
+interface WorkflowsToolbarProps {
+    selectedWorkflows: {[index: string]: Workflow};
+    loadWorkflows: () => void;
+    isDisabled: Actions.ActionDisabled;
+}
+
+interface WorkflowGroupAction {
```
Comment (simster7) [inline ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:16]: It seems like this is just extended from `WorkflowAction`. Could we just extend it? https://www.typescriptlang.org/docs/handbook/interfaces.html#extending-interfaces

File: ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    public render() {
+        return (
+            <Consumer>
+                {ctx => (
+                    <div className='workflows-toolbar'>
+                        <div className='workflows-toolbar__count'>
+                            {this.noneSelected() ? 'No' : this.getNumberSelected()}
+                            &nbsp;workflow{this.getNumberSelected() === 1 ? '' : 's'} selected
```
Comment (simster7) [inline ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:41]: Love the attention to detail(s)!

File: ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+                        <div className='workflows-toolbar__actions'>{this.renderActions(ctx)}</div>
+                    </div>
+                )}
+            </Consumer>
+        );
+    }
+
+    private noneSelected(): boolean {
+        return Object.keys(this.props.selectedWorkflows).length < 1;
+    }
```
Comment (simster7) [inline ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:52]: Why is this necessary? Couldn't we just do `getNumberSelected() == 0`?

File: ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+    private getNumberSelected(): number {
+        return Object.keys(this.props.selectedWorkflows).length;
```
Comment (simster7) [inline ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:55]: I could be wrong, but doesn't
```suggestion
        return this.props.selectedWorkflows.length;
```
work?

File: ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+                });
+        }
+    }
+
+    private confirmAction(title: string): void {
+        if (!confirm(`Are you sure you want to ${title} all selected workflows?`)) {
+            return;
+        }
+        return;
+    }
```
Comment (simster7) [inline ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:80]: Same here regarding the `if` statement

File: ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+
+    private getHandleErrorFunction(title: string): () => void {
+        return () => {
+            this.props.loadWorkflows();
+            this.appContext.apis.notifications.show({
+                content: `Unable to ${title} workflow`,
+                type: NotificationType.Error
+            });
+        };
+    }
```
Comment (simster7) [inline ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:90]: This could also be inlined

File: ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+                disabled: disabled[actionName],
+                action: () => this.performActionOnSelectedWorkflows(ctx, action.title, action.action),
+                className: action.title
+            };
+        });
+    }
+
+    private renderActions(ctx: any): JSX.Element[] {
+        const actionButtons = [];
+        for (const action of this.getActions(ctx)) {
```
Comment (simster7) [inline ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:109]: I feel like we don't need `getActions` and can do the whole thing in one loop

File: ui/src/app/shared/workflow-actions.ts (hunk 2 above)
Comment (rbreeze) [reply ui/src/app/shared/workflow-actions.ts:13]: I made this `any` because of the `WorkflowDeleteResponse` type used here in `workflows-service.ts`: 

```js
public delete(name: string, namespace: string): Promise<WorkflowDeleteResponse> {...
```

For some reason, the delete function is the only one with its own Response type. I'll look into making a generic `WorkflowActionResponse` type to get rid of that `any`

File: ui/src/app/workflows/components/workflow-details/workflow-details.tsx (hunk 4 above)
Comment (rbreeze) [reply ui/src/app/workflows/components/workflow-details/workflow-details.tsx:205]: I'm fairly confident that as of ES2015 the order of object props can be guaranteed, and as of ES6 `Object.keys` relies on the underlying method that they use to guarantee the order. Here's the relevant portion of the spec: 
https://tc39.es/ecma262/#sec-ordinaryownpropertykeys

File: ui/src/app/workflows/components/workflow-details/workflow-details.tsx (hunk 5 above)
Comment (rbreeze) [reply ui/src/app/workflows/components/workflow-details/workflow-details.tsx:208]: yeah I wasn't wild about that either. Capital case is a good solution and `*.lower()` is certainly cleaner, I'll do that.

File: ui/src/app/workflows/components/workflows-list/workflows-list.tsx (hunk 9 above)
Comment (rbreeze) [reply ui/src/app/workflows/components/workflows-list/workflows-list.tsx:78]: I added this because when changing filters, labels or pagination, there was a memory leak due to workflows that were out of view still being held in the `selectedWorkflows` state. I can't remember what the exact console error was but I can find out

File: ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx (hunk 12 above)
Comment (rbreeze) [reply ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:16]: Yeah, definitely. Wasn't aware of that syntax

File: ui/src/app/workflows/components/workflow-details/workflow-details.tsx (hunk 3 above)
Comment (rbreeze) [reply ui/src/app/workflows/components/workflow-details/workflow-details.tsx:185]: I think this is correct, I tried it without the `if` and it didn't work (the confirm dialogue didn't pop up)

File: ui/src/app/workflows/components/workflows-list/workflows-list.tsx (hunk 11 above)
Comment (rbreeze) [reply ui/src/app/workflows/components/workflows-list/workflows-list.tsx:188]: Yep! Changed

File: ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx (hunk 13 above)
Comment (rbreeze) [reply ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:41]: Thanks!

File: ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx (hunk 14 above)
Comment (rbreeze) [reply ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:52]: Yeah, good point. Changed

File: ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx (hunk 15 above)
Comment (rbreeze) [reply ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:55]: I tried it, it doesn't - since `selectedWorkflows` is an object rather than an array it doesn't have a `length` property

File: ui/src/app/shared/workflow-actions.ts (hunk 2 above)
Comment (simster7) [reply ui/src/app/shared/workflow-actions.ts:13]: I see. If we don't use the result of the promise anywhere then this should be fine

File: ui/src/app/shared/workflow-actions.ts (hunk 7 above)
Comment (simster7) [reply ui/src/app/shared/workflow-actions.ts:15]: It doesn't seem like this was changed?

File: ui/src/app/workflows/components/workflow-details/workflow-details.tsx (hunk 4 above)
Comment (simster7) [reply ui/src/app/workflows/components/workflow-details/workflow-details.tsx:200]: Great! Thanks

File: ui/src/app/workflows/components/workflow-details/workflow-details.tsx (hunk 5 above)
Comment (simster7) [reply ui/src/app/workflows/components/workflow-details/workflow-details.tsx:203]: It doesn't seem like this was changed?

File: ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx (hunk 15 above)
Comment (simster7) [reply ui/src/app/workflows/components/workflows-toolbar/workflows-toolbar.tsx:49]: Great

File: ui/src/app/workflows/components/workflow-details/workflow-details.tsx (hunk 3 above)
Comment (simster7) [reply ui/src/app/workflows/components/workflow-details/workflow-details.tsx]: I meant more that it doesn't matter if `confirm` returns `true` or `false`, the result doesn't chagne

**Comment by rbreeze** [comment]
Ah interesting, will test on Firefox and fix. Might be a z-index issue. 

I was thinking along the same lines but was undecided, that makes the decision easy... will change both

File: ui/src/app/shared/workflow-actions.ts [hunk 19]
```diff
@@ -0,0 +1,64 @@
+import {NodePhase, Workflow} from '../../models';
+import {services} from './services';
+import {Utils} from './utils';
+
+export type ActionDisabled = {
+    [action in WorkflowActionName]: boolean;
+};
+
+export type WorkflowActionName = 'RETRY' | 'RESUBMIT' | 'SUSPEND' | 'RESUME' | 'STOP' | 'TERMINATE' | 'DELETE';
```
Comment (alexec) [inline ui/src/app/shared/workflow-actions.ts:9]: the actions map to API which use lowercase names?

Comment (alexec) [reply ui/src/app/shared/workflow-actions.ts:9]: also - I think you should use the term `operation` rather than `action`

File: ui/src/app/shared/workflow-actions.ts [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+
+export interface WorkflowAction {
+    title: string;
+    action: () => Promise<any>;
+    iconClassName: string;
+    disabled: (wf: Workflow) => boolean;
+}
+
+export const WorkflowActions = {
```
Comment (alexec) [inline ui/src/app/shared/workflow-actions.ts:18]: I like this

File: ui/src/app/shared/workflow-actions.ts [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+        disabled: (wf: Workflow) => !Utils.isWorkflowSuspended(wf),
+        action: services.workflows.stop
+    },
+    TERMINATE: {
+        title: 'TERMINATE',
+        iconClassName: 'fa fa-times-circle',
+        disabled: (wf: Workflow) => !Utils.isWorkflowSuspended(wf),
+        action: services.workflows.terminate
+    },
+    DELETE: {
```
Comment (alexec) [inline ui/src/app/shared/workflow-actions.ts:58]: can we make sure that we get an "are you sure" confirmation for delete?

File: ui/src/app/shared/workflow-actions.ts (hunk 19 above)
Comment (rbreeze) [reply ui/src/app/shared/workflow-actions.ts:9]: I had them as lowercase but changed to upper for labeling purposes. Simon pointed out that the way I was initially converting lowercase to a label (`string.charAt(0).toUpperCase() + string.slice(1);`) was clunky and just having them as uppercase would be easier. Thinking about it now though, it wouldn't be clunky to have them lowercase and do `string.toUpperCase()` to convert. The union type is never used to actually query the API here, so would the change to lowercase be for consistency? 

Operation sounds good, I'll change that.

File: ui/src/app/shared/workflow-actions.ts (hunk 20 above)
Comment (rbreeze) [reply ui/src/app/shared/workflow-actions.ts:18]: thanks!

File: ui/src/app/shared/workflow-actions.ts (hunk 21 above)
Comment (rbreeze) [reply ui/src/app/shared/workflow-actions.ts:58]: In the workflows toolbar and the workflows details, there's a confirmation before deleting - in fact for the workflows toolbar, there's a confirmation before performing any batch action. Would it be better to have the confirmation happen in `workflow-actions.ts`?

File: ui/src/app/shared/services/workflows-service.ts [hunk 22]
```diff
@@ -69,7 +70,8 @@ export class WorkflowsService {
             'result.object.status.phase',
             'result.object.status.startedAt',
             'result.type',
-            'result.object.metadata.labels'
+            'result.object.metadata.labels',
+            'result.object.spec'
```
Comment (alexec) [inline ui/src/app/shared/services/workflows-service.ts:74]: do you want the whole spec here? could be a lot of data

File: ui/src/app/shared/workflow-operations.ts [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+import {services} from './services';
+import {Utils} from './utils';
+
+export type OperationDisabled = {
+    [action in WorkflowOperationName]: boolean;
+};
+
+export type WorkflowOperationName = 'RETRY' | 'RESUBMIT' | 'SUSPEND' | 'RESUME' | 'STOP' | 'TERMINATE' | 'DELETE';
+
+export interface WorkflowAction {
```
Comment (alexec) [inline ui/src/app/shared/workflow-operations.ts:11]: `WorkflowOperation`?

File: ui/src/app/workflows/components/workflow-details/workflow-details.tsx [hunk 24]
```diff
@@ -14,7 +14,7 @@ import {CostOptimisationNudge} from '../../../shared/components/cost-optimisatio
 import {Loading} from '../../../shared/components/loading';
 import {hasWarningConditionBadge} from '../../../shared/conditions-panel';
 import {Consumer, ContextApis} from '../../../shared/context';
-import {Utils} from '../../../shared/utils';
+import * as Actions from '../../../shared/workflow-operations';
```
Comment (alexec) [inline ui/src/app/workflows/components/workflow-details/workflow-details.tsx:17]: `Operations`?

**Comment by rbreeze** [comment]
of course, here are a couple: 
Screen Shot 2020-06-22 at 12 36 31 PM
Screen Shot 2020-06-22 at 12 43 43 PM

**Comment by hadim** [comment]
I can't see this feature in 2.9.1. Is that planned for 2.10 or do we need to do something to enable it?
